### PR TITLE
OMG-25427 - Set SHIPPO-ACCOUNT-ID header

### PR DESCRIPTION
**Pull request by slavick**
[OMG-25427](https://omgear.atlassian.net/browse/OMG-25427)

Update to support the `SHIPPO-ACCOUNT-ID` header required for Shippo's new sub-account system.

Related PR:
* https://github.com/OrderMyGear/shipping-service/pull/223

[OMG-25427]: https://omgear.atlassian.net/browse/OMG-25427?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -131,6 +127,21 @@ func (c *Client) createRequest(method, url string, bodyObject interface{}) (req
...
+			subAcctID = v.ShippoSubAccountID
+		case models.ShipmentInput:
+			subAcctID = v.ShippoSubAccountID
+		case models.ShippoSubAccount:
+			subAcctID = v.ShippoSubAccountID
+		case models.TrackingStatusInput:
+			subAcctID = v.ShippoSubAccountID
+		case models.TransactionInput:
+			subAcctID = v.ShippoSubAccountID
+		}
```
Comment (slavick) [inline client/client.go]: I know I should be able to combine all the types in a single `case` statement to eliminate the duplicated lines, but the compiler complains.

Screenshot 2024-09-11 at 9 28 55 AM

**Review by ilyutov — COMMENTED** [review]
Why don't we add a separate "headers" argument to `Client.do` instead of changing every struct definition?

**Comment by slavick** [comment]
> Why don't we add a separate "headers" argument to `Client.do` instead of changing every struct definition?

Good idea! I made the changes in this PR and am in the process of updating the shipping-service PR.

**Review by ilyutov — APPROVED** [review]
👍

File: client/client.go [hunk 2]
```diff
@@ -149,6 +142,14 @@ func (c *Client) createRequest(method, url string, bodyObject interface{}) (req
 	req.Header.Set("Connection", "close")
 	req.Close = true
 
+	// add any passed in headers
+	if headers != nil {
+		for k, v := range headers {
+			c.logPrintf("Client.createRequest() setting header: key=%q, value=%q", k, v)
```
Comment (permanuno) [inline client/client.go]: might want to relocate the `c.logPrintf` to the `defer func` above, looks like logging is only done if the logger isn't `nil`

**Review by permanuno — CHANGES_REQUESTED** [review]

File: client/client.go (hunk 2 above)
Comment (slavick) [reply client/client.go]: Done, but `logPrintf` also checks if logger is nil.

**Review by permanuno — APPROVED** [review]

File: client/tracking_status.go [hunk 3]
```diff
@@ -8,7 +8,7 @@ import (
 )
 
 // GetTrackingUpdate requests the tracking status of a shipment.
-func (c *Client) GetTrackingUpdate(carrier, trackingNumber string) (*models.TrackingStatus, error) {
+func (c *Client) GetTrackingUpdate(carrier, trackingNumber string, shippoSubAccountID string) (*models.TrackingStatus, error) {
```
Comment (permanuno) [inline client/tracking_status.go]: i don't think we'll need subaccount here, there's a carrier but it's `usps` not the carrier id: https://docs.goshippo.com/shippoapi/public-api/#operation/GetTrack

**Review by permanuno — CHANGES_REQUESTED** [review]
i think `GetTrackingUpdate` needs an update.

**Review by permanuno — APPROVED** [review]

**Comment by slavick** [comment]
> Why don't we add a separate "headers" argument to `Client.do` instead of changing every struct definition?

@ilyutov This is the strategy we went with. Please review when you have a chance.

**Review by ilyutov — APPROVED** [review]
LGTM

**Pull request merged**